[AArch64] Reduce the cost of repeated sub-shuffle

davemgreen opened the pull request:
Given a larger-than-legal shuffle we will split into multiple sub-parts. This adds a check to the computed costs of sub-shuffles so that repeated sequences are not accounted for multiple times. This especially reduces the cost of broadcasts/splats.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -5514,15 +5516,25 @@ InstructionCost AArch64TTIImpl::getShuffleCost(
         else
           NMask.push_back(MaskElt % LTNumElts);
       }
+      // Check if we have already generated this sub-shuffle, which means we
+      // will have already generated the output. For example a <16 x i32> splat
+      // will be the same sub-splat 4 times, which only needs to be generated
+      // once and reused.
+      auto Result =
+          PreviousCosts.insert({std::make_tuple(Source1, Source2, NMask), 0});
+      if (!Result.second)
+        continue;
```
It wasn't immediately clear that `Result.second` isn't the map's entry's value and is actually whether the entry was inserted or not. Could you explain that a little in the comment?

SamTebbs33 reviewed (approved):
LGTM with one suggestion.

c-rhodes reviewed (approved):
also LGTM cheers

davemgreen committed:
[AArch64] Reduce the cost of repeated shuffle costs

Given a larger-than-legal shuffle we will split into multiple sub-parts. This
adds a check to the computed costs of sub-shuffles so that repeated sequences
are not accounted for multiple times. This especially reduces the cost of
broadcasts/splats.

qinkunbao commented:
Hi, could you take a look at the buildbot failure?

https://lab.llvm.org/buildbot/#/builders/94/builds/7184

davemgreen commented:
Hi - I've pushed ff78d233c0c40f94ef922a4ea1b88e02bd126040 but am not sure if it will fix it. If it persists then please (you or someone else who is still up) feel free to revert. Otherwise I will check in the morning. Thanks.
